files/custom.lua

```diff
@@ -135,27 +143,77 @@ function mysplit(str, sep)
   string.gsub(str, "[^" .. sep .. "]+", func)
   return res
 end
---function mysplit(inputstr, sep)
---        if sep == nil then
---                sep = ":"
---        end
---        local t={}
---        for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
---                table.insert(t, str)
---        end
---        return t
---end
--- function auth_on_publish(pub)
---   return next
--- end
---    return true
---end
---pool = "auth_custom"
---config = {
---    pool_id = pool
---}
 
---custom.ensure_pool(config)
+function equals(o1, o2)
+  if o1 == o2 then return true end
+  local o1Type = type(o1)
+  local o2Type = type(o2)
+  if o1Type ~= o2Type then return false end
+  if o1Type ~= 'table' then return false end
+
+  local keySet = {}
+
+  for key1, value1 in pairs(o1) do
+      local value2 = o2[key1]
+      if value2 == nil or equals(value1, value2) == false then
+          return false
+      end
+      keySet[key1] = true
+  end
+
+  for key2, _ in pairs(o2) do
+      if not keySet[key2] then return false end
+  end
+  return true
+end
+
+function cmd(cmd, expected_result)
+  result = redis.cmd("redis_dev", cmd)
+  return equals(result, expected_result)
+end
+
+-- redis keys will be deleted from chat server when stream terminates
+function auth_on_publish(pub)
+  local stream_id = string.match(pub.topic, "chat/(.+)/message")
+  if stream_id then
+    local mod_payload, count = string.gsub(pub.payload, "\\n*", "")
+    local decoded_payload = json.decode(mod_payload)
+    local user_uid
+
+    if not decoded_payload or string.len(decoded_payload["message"]) > 240 or not decoded_payload["profile"] then
+      return false
+    end
+
+    user_uid = decoded_payload["profile"]["uid"]
+
+    -- verify commands
+    if decoded_payload["profile"]["is_streamer"] then
+      -- if profile_uid is streamer check for ban in message
+      if string.match(decoded_payload["message"], "^/ban ") then
+        local tagged_profile = decoded_payload["tagged_profile"]
+        if not tagged_profile then
+          return {topic = string.format("chat/%s/command", stream_id), payload = mod_payload, throttle = 1000 }
+        end
+        cmd(string.format("hmset ban:%s %s %s", stream_id, tagged_profile["uid"], "1"), true)
+        return {topic = string.format("chat/%s/command", stream_id), payload = mod_payload, throttle = 1000 }
+      end
+    else
+      -- if not check if profile uid is banned if yes return false
+      if cmd(string.format("hget ban:%s %s", stream_id, user_uid), "1") then
```

Comment:
will return `true` when user_uid present in the hash set?

Reply:
It will forward message to chat/+/command. Message won't be published to chat/+/message